### ISSUE TITLE
Call parent class serializer for collection objects (at least)

### ISSUE THREAD

**Issue by lifflander** [DARMA-tasking/vt#432]
Automatically have the system invoke the parent class serializer so the user doesn't make the mistake of not calling it.

**Comment by lifflander** [comment]
Based on a conversation with @PhilMiller, we are going to implement a general mechanism in checkpoint `template <typename T> struct inherit : T {}` that does auto-dispatch to base classes. We will use this mechanism in VT for `vt::Collection<T,IndexT>`

**Comment by PhilMiller** [comment]
Corresponds to https://github.com/DARMA-tasking/checkpoint/issues/70

**Comment by PhilMiller** [comment]
@lifflander In code like this:

`struct VecMsg : vt::CollectionMessage<LinearPb2DJacobi>`

Does `VecMsg::serialize()` have to call anything for its base class? Is the runtime doing something magical there, like always byte-copying `sizeof(msg)` in addition to serialization?

**Comment by PhilMiller** [comment]
`struct VirtualMsg` in `src/vt/vrt/context/context_vrtmessage.h` presents an interesting case. It is marked as byte-serializable, but supports descendent classes that may not be.

I think it may be ok to make that just a straight `serialize()` method for the moment, since it should be pretty easy to then dispatch the 'base class is byte-serializable' case separate from the child, or let them combine.

**Comment by PhilMiller** [comment]
There's one example that doesn't compile out of all of these changes: `hello_world_vc_remote.cc`. It produces the following errors, which may perhaps be real?

```
In instantiation of ‘void serdes::Base<T, SK>::_serdes_internal_serialize(SerializerT&) [with SerializerT = serdes::Unpacker;                                                                                                                                                 
Tcalled = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, MyVC>;                                                                                                                                                                                            
T       = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, vt::vrt::VirtualContext>;                                                                                                                                                                         
serdes::SerializeKind SK = (serdes::SerializeKind)1]’:

In instantiation of ‘void serdes::Base<T, SK>::_serdes_internal_serialize(SerializerT&) [with SerializerT = serdes::Sizer;                                                                                                                                                    
Tcalled = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, MyVC>;                                                                                                                                                                                            
T       = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, vt::vrt::VirtualContext>;                                                                                                                                                                         
serdes::SerializeKind SK = (serdes::SerializeKind)1]’

In instantiation of ‘void serdes::Base<T, SK>::_serdes_internal_serialize(SerializerT&) [with SerializerT = serdes::PackerBuffer<serdes::ManagedBuffer>;                                                                                                                      
Tcalled = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, MyVC>;                                                                                                                                                                                            
T       = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, vt::vrt::VirtualContext>;                                                                                                                                                                         
serdes::SerializeKind SK = (serdes::SerializeKind)1]

In instantiation of ‘void serdes::Base<T, SK>::_serdes_internal_serialize(SerializerT&) [with SerializerT = serdes::PackerBuffer<serdes::UserBuffer>;                                                                                                                         
Tcalled = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, MyVC>;                                                                                                                                                                                            
T       = vt::vrt::VrtConstructMsg<vt::vrt::RemoteVrtInfo, std::tuple<int>, vt::vrt::VirtualContext>;                                                                                                                                                                         
serdes::SerializeKind SK = (serdes::SerializeKind)1]’
```

There's this in the definition: `struct MyVC : vt::vrt::VirtualContext`. So they're related, but not exactly the same type.

Lack of GTest means that I don't have the tests being compiled at the moment.

**Comment by PhilMiller** [comment]
VT is probably the worst case for the pain of the base class serialization automation. We'll want to consider whether the design is good enough for use here and elsewhere, whether it's overkill, or there's something we can add to it to reduce some of this pain. 

I'll get started on seeing what needs to be done in/on EMPIRE, but I suspect that will all be a lot simpler than what we see here. Any inheritance there should actually be simple layering, without hugely different intended serialization semantics between parent and child.

**Comment by PhilMiller** [comment]
And I see test failure in `examples/ctx_callback` that I actually need to investigate and resolve. The callbacks were the most involved part of the change, so it's unsurprising something didn't work there.

**Comment by PhilMiller** [comment]
`jacobi2d` hangs at all process counts, indicating I likely broke something there as well.

**Comment by PhilMiller** [comment]
> And I see test failure in `examples/ctx_callback` that I actually need to investigate and resolve. The callbacks were the most involved part of the change, so it's unsurprising something didn't work there.

The only change there was this:

```
-struct DataMsg : Message {
+struct DataMsg : Message, serdes::Base<DataMsg> {
```

I can't use `Inherit<DataMsg, Message>` because Message omits a `serialize` method entirely and is just byte-serialized. I believe we want to maintain that.

Perhaps this indicates that we need to add a byte serialization trait alongside the `HasSerialize, EmptySerialize` options that are presented now. The design of this may be a bit finicky if we want to allow for some classes in an inheritance hierarchy to be byte serialized but not others, involving interesting calculations around `sizeof`. Consider, for instance, an object whose `sizeof` is rounded up from 19 bytes of members to 24 bytes for alignment. A class derived from that would pack members into the excess 5 bytes as appropriate (e.g. an `int` at byte 20), so we can't just start copying for a derived class at the `sizeof` its base class, for instance.

I want to suggest a more conservative approach would be to byte-serialize the first N bytes of an object where N is chosen based on the longest base class that wants byte serialization, or even just the most-derived class. Then we can figure that the explicit serializers will fill in the members for which byte serialization fails. However, that doesn't actually work, because we need to precisely deserialize the bytes that make up the members potentially interleaved with construction, to avoid overwriting at the wrong times.

**Comment by PhilMiller** [comment]
Editorial remark: stringently enforcing that the class being serialized matches the serializer being call has made it really easy to pursue this task relatively mechanically. I just need to read off the two classes appearing in the instantiation as shown above, and walk the up the hierarchy of definitions between them.

**Comment by PhilMiller** [comment]
Regarding byte serialization: there's a somewhat natural seeming restriction that would substantially simplify things, of requiring that if a derived class calls for byte serialization, its base class must as well (and this applies recursively). I believe that constraint was already there, though not syntactically codified.

**Comment by lifflander** [comment]
@PhilMiller Sorry for the delay, I missed these comments.

I think that's actually a really good property to encode:  if a derived class is byte serialized, its base must also as well... recursively down the inheritance graph.

For `struct VecMsg : vt::CollectionMessage<LinearPb2DJacobi>`, the system in indeed doing something magical. The reason this was implemented was because of how often I forgot to call the `CollectionMessage<T>` serializer. Also, I needed to not have a `CollectionMessage<T>::serialize()` so it was byte serializable when the user called for that.

Thus, I added a trait in checkpoint `hasSplitSerialize<U>`. This means that it has `CollectionMessage<T>::serializeThis()` and `CollectionMessage<T>::serializeParent()`. If a class is being serialized already, the system checks for both of those, and then calls them when serialize is not available. This enables the trait that decides whether to serialize or not based on whether the method exists to work properly, and for it to automatically be called.
